suprb/suprb.py

```diff
@@ -186,8 +193,14 @@ def _discover_rules(self, X: np.ndarray, y: np.ndarray, n_rules: int):
         if not self.pool_:
             warnings.warn(
                 "The population is empty, even after generating rules. "
-                "Solution optimization will be skipped.",
+                "The run will be terminated.",
```

Comment:
We don't want to terminate if we can't generate a rules (at least not immediately). My suggestion would be to retry generating a new rule for a certain number of times (with a warning saying that we needed multiple iterations to create a new rule) and only if that also fails we terminate the run. 

Also it is very confusing if the `self._discover_rules` suddenly has a return value, that is not intuitive

Reply:
I don't fully agree. Running multiple tries could work, but we should never terminate. We would at least want to add an overgeneral rule. However (!), I posit that having no rules in the pool should not be an issue. The fallback prediction for unmatched examples is the mean of the data, so we do cover the case of a ruleless model already. We just need to give that one a sensible fitness. Which we can. And should do

Reply:
Added the proposed mechanism of repeatedly attempting to find rules, if none could be found (Currently max. 4 attempts). If that also fails then one overgeneral rule is added into pool and the run is continued. Having no rules in the pool at the time when Solution Composition is called causes a error:

 File "C:\Users\anton\PycharmProjects\suprb\suprb\suprb.py", line 227, in _compose_solution
    self.solution_composition_.optimize(X, y)
  File "C:\Users\anton\PycharmProjects\suprb\suprb\optimizer\solution\base.py", line 109, in optimize
    self.archive(self.population_)
  File "C:\Users\anton\PycharmProjects\suprb\suprb\optimizer\solution\archive.py", line 36, in __call__
    best = max(new_population, key=lambda i: i.fitness_)
  File "C:\Users\anton\PycharmProjects\suprb\suprb\optimizer\solution\archive.py", line 36, in <lambda>
    best = max(new_population, key=lambda i: i.fitness_)
AttributeError: 'Solution' object has no attribute 'fitness_'

  warnings.warn(some_fits_failed_message, FitFailedWarning)

steming from the fact that if the pool is empty, the population is never actually fitted causing the attributes of the individuals to not yet be initialized:
https://github.com/heidmic/suprb/blob/80e72e1af49cf9e9635adfb50d6d05f83f2ab3f2/suprb/optimizer/solution/base.py#L102-L105

Im not certain as to how best to approach this issue. Adding a synthetic rule into the pool manages to prevent this issue from occuring but most likely isn't the ideal way.